File: src/main/java/net/wapic/wpcmod/events/PacketEvents.java
package net.wapic.wpcmod.events;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.client.world.ClientWorld;
import net.minecraft.network.packet.s2c.play.ParticleS2CPacket;
import net.minecraft.network.packet.s2c.play.PlaySoundS2CPacket;

public interface PacketEvents {
    Event<ParticleEvent> PARTICLE = EventFactory.createArrayBacked(ParticleEvent.class, (listeners) -> (packet, world) -> {
            for (ParticleEvent listener : listeners) {
                listener.onParticle(packet, world);
            }
    });

    Event<SoundEvent> PLAY_SOUND = EventFactory.createArrayBacked(SoundEvent.class, (listeners) -> (packet, world) -> {
        for (SoundEvent listener : listeners) {
            listener.onPlaySound(packet, world);
        }
    });

    interface ParticleEvent {
        void onParticle(ParticleS2CPacket packet, ClientWorld world);
    }

    interface SoundEvent {
        void onPlaySound(PlaySoundS2CPacket packet, ClientWorld world);
    }
}

File: src/main/kotlin/net/wapic/wpcmod/events/PacketEvents.kt
package net.wapic.wpcmod.events

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.network.PacketListener
import net.minecraft.network.protocol.Packet
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo

object PacketEvents {

	@JvmField
	val RECEIVE: Event<PacketReceive> = EventFactory.createArrayBacked(PacketReceive::class.java) { listeners ->
		PacketReceive { packet ->
			for (listener in listeners) {
				listener.onPacketReceive(packet)
			}
		}
	}

	fun interface PacketReceive {
		fun onPacketReceive(packet: Packet<*>)
	}

	@JvmField
	val SEND_BEFORE: Event<PacketSendBefore> =
		EventFactory.createArrayBacked(PacketSendBefore::class.java) { listeners ->
			PacketSendBefore { packet, callbackInfo ->
			for (listener in listeners) {
				listener.onPacketSendBefore(packet, callbackInfo)
			}
		}
	}

	fun interface PacketSendBefore {
		fun onPacketSendBefore(packet: Packet<out PacketListener>, callbackInfo: CallbackInfo)
	}

	@JvmField
	val SEND_AFTER: Event<PacketSendAfter> = EventFactory.createArrayBacked(PacketSendAfter::class.java) { listeners ->
		PacketSendAfter { packet, callbackInfo ->
			for (listener in listeners) {
				listener.onPacketSendAfter(packet, callbackInfo)
			}
		}
	}

	fun interface PacketSendAfter {
		fun onPacketSendAfter(packet: Packet<out PacketListener>, callbackInfo: CallbackInfo)
	}
}